src/mrpro/operators/WaveletOp.py

```diff
@@ -0,0 +1,350 @@
+            forward operator. The adjoint operator will raise an error.
+        dim
+            Dimensions (axes) where wavelets are calculated
+        wavelet_name
+            Name of wavelets
+        level
+            Highest wavelet level. If set to None, the highest possible level is calculated based on the domain shape.
+
+        Raises
+        ------
+        ValueError
+            If wavelets are calculated for more than three dimensions.
+        ValueError
+            If wavelet dimensions and domain shape do not match.
+        NotImplementedError
+            If the domain shape is odd. Adjoint will lead to the wrong domain shape.
+        """
+        super().__init__()
+        self.domain_shape = domain_shape
+        self.wavelet_name = wavelet_name
+        self.dim = dim
+        self.level = level
+
+        # number of wavelet directions
+        if len(dim) == 1:
+            self.n_wavelet_directions = 1
+        elif len(dim) == 2:
+            self.n_wavelet_directions = 3
+        elif len(dim) == 3:
+            self.n_wavelet_directions = 7
+        else:
+            raise ValueError('Only 1D, 2D and 3D wavelet transforms are supported.')
+
+        if domain_shape is not None:
+            if len(dim) != len(domain_shape):
+                raise ValueError(
+                    'Number of dimensions along which the wavelet transform should be calculated needs to',
+                    'be the same as the domain shape',
+                )
+
+            if any(d % 2 for d in domain_shape):
+                raise NotImplementedError('ptwt only supports wavelet transforms for even number of samples.')
+
+            # size of wavelets
+            wavelet_length = torch.as_tensor((Wavelet(wavelet_name).dec_len,) * len(domain_shape))
+
+            # calculate shape of wavelet coefficients at each level
+            current_shape = torch.as_tensor(domain_shape)
+
+            if _check_level(domain_shape, wavelet_length, level) == 0:
+                self.coefficients_shape = [current_shape]
+                self.coefficients_padding = [(0, 0)] * len(current_shape)
+            else:
+                self.coefficients_shape = []
+                self.coefficients_padding = []
+                for _ in range(_check_level(domain_shape, wavelet_length, level)):
+                    # Add padding
+                    for ind in range(len(current_shape)):
+                        padl, padr = _get_pad(current_shape[ind], wavelet_length[ind])
+                        current_shape[ind] += padl + padr
+                    current_shape = torch.floor((current_shape - (wavelet_length - 1) - 1) / 2 + 1).to(
+                        dtype=torch.int64
+                    )
+                    self.coefficients_shape.extend([current_shape.clone()] * self.n_wavelet_directions)
+                    self.coefficients_padding.extend([(padl.clone(), padr.clone())] * self.n_wavelet_directions)
```

Comment:
padl and padr will always be equal for even images, and equal `(2 * filt_len - 3) // 2`

Reply:
yes, but it does not depend on the image but the current shape of the current level and this can be odd. In this case the padding is (2 * filt_len - 3) // 2 + 1. I could of course include this even/odd check in our code but IMO it makes more sense to call their function

Reply:
1. it is a private function
2. it is a lot easier to understand that the shape will be:
`(current_shape/2).ceil()+wavelet_length//2-1`
and this hole section can just be collapsed.

Reply:
why is current_shape a tensor, why is padl and padr tensors?
_getpad has typehints int

Reply:
(and why the clones..?)

Reply:
and is      
`self.coefficients_padding.extend([(padl.clone(), padr.clone())] * self.n_wavelet_directions)`
correct? 
padl and padr are updated in the loop above, this should only be the last value? I am really confused by this whole section...

Reply:
the clones are needed because extends does only add a view to the object and hence padl and padr would be overwritten in the loop

Reply:
tensors are there for the math or am I missing something?

Reply:
mazbe either do everything as tensors in this file,
the whole current shape logic couldbe written as (including the padding)
(current_shape/2).ceil()+wavelet_length//2-1

or use the _get_pad function and iterate over all values.

the mix of vectorized tensor logic and the loop seems complicated.

Reply:
- why is coefficients padding not just a list of ints?
-  why is self.coefficients_shape not a list of ints?

I would either do all math in this section on the tensors, then convert to lists of ints to store (so no loop over the elements of the tensor) or just do everything as a loop

Reply:
also... is coefficients_padding ever used?

Reply:
if not, see suggestion below